tests/testthat/test-image_geom.R

```diff
@@ -0,0 +1,84 @@
+  "type",  "group2",       "Type of engine",           "image",      "palette2",  lst(directory = "test_images/", extension = "png")
+)
+
+data_path <- mtcars %>%
+  rownames_to_column("id") %>%
+  arrange(desc(mpg)) %>%
+  head(20)
+data_path$type <- c(rep("test_images/ice.png", 10), rep("test_images/electric.png", 10))
+
+column_info_data_path <- tribble(
+  ~id,     ~group,         ~name,                      ~geom,        ~palette,    ~options,
+  "id",    NA,             "",                         "text",       NA,          list(hjust = 0, width = 6),
+  "mpg",   "overall",      "Miles / gallon",           "bar",        "palette1",  list(width = 4, legend = FALSE),
+  "cyl",   "overall",      "Number of cylinders",      "bar",        "palette2",  list(width = 4, legend = FALSE),
+  "disp",  "group1",       "Displacement (cu.in.)",    "funkyrect",  "palette1",  lst(),
+  "hp",    "group1",       "Gross horsepower",         "funkyrect",  "palette1",  lst(),
+  "drat",  "group1",       "Rear axle ratio",          "funkyrect",  "palette1",  lst(),
+  "wt",    "group1",       "Weight (1000 lbs)",        "funkyrect",  "palette1",  lst(),
+  "qsec",  "group2",       "1/4 mile time",            "circle",     "palette2",  lst(),
+  "vs",    "group2",       "Engine",                   "circle",     "palette2",  lst(),
+  "am",    "group2",       "Transmission",             "circle",     "palette2",  lst(),
+  "gear",  "group2",       "# Forward gears",          "circle",     "palette2",  lst(),
+  "carb",  "group2",       "# Carburetors",            "circle",     "palette2",  lst(),
+  "type",  "group2",       "Type of engine",           "image",      "palette2",  lst()
+)
+
+
+row_info <- data %>% transmute(id, group = "test")
+row_groups <- tibble(Group = "Test", group = "test")
+
+column_groups <- tribble( # tribble_start
+  ~Category,  ~group,         ~palette,
+  "Overall",  "overall",      "overall",
+  "Group 1",  "group1",       "palette1",
+  "Group 2",  "group2",       "palette2"
+) # tribble_end
+
+palettes <- tribble(
+  ~palette,             ~colours,
+  "overall",            grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(9, "Greys")[-1]))(101),
+  "palette1",           grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(9, "Blues") %>% c("#011636")))(101),
+  "palette2",           grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(9, "Reds")[-8:-9]))(101)
+)
+g <- funky_heatmap(
+  data = data,
+  column_info = column_info,
+  column_groups = column_groups,
+  row_info = row_info,
+  row_groups = row_groups,
+  palettes = palettes,
+  expand = list(xmax = 4)
+)
+ggsave("test3.pdf", g, device = "pdf")
+
+g_path <- funky_heatmap(
+  data = data_path,
+  column_info = column_info_data_path,
+  column_groups = column_groups,
+  row_info = row_info,
+  row_groups = row_groups,
+  palettes = palettes,
+  expand = list(xmax = 4)
+)
+ggsave("test_data_path.pdf", g, device = "pdf")
```

Comment:
No need to write the pdfs to file, IMO. Also, this file does not contain any actual tests.

There's also a lot of additional code which is not really relevant to the test (though perhaps edge cases can occur with all of the extra data, who knows).

I'd suggest:

```suggestion
test_that("image geom with full paths", {
  data <- mtcars %>%
    rownames_to_column("id") %>%
    arrange(desc(mpg)) %>%
    head(20) %>%
    mutate(type = c(rep("test_images/ice.png", 10), rep("test_images/electric.png", 10)))

  column_info <- tribble(
    ~id,     ~group,         ~name,                      ~geom,        ~palette,    ~options,
    "id",    NA,             "",                         "text",       NA,          list(hjust = 0, width = 6),
    "mpg",   NA,             "Miles / gallon",           "bar",        NA,          list(width = 4, legend = FALSE),
    "type",  NA,             "Type of engine",           "image",      NA,          lst(directory = "test_images/", extension = "png")
  )
  
  g <- funky_heatmap(
    data = data,
    column_info = column_info,
    expand = list(xmax = 4)
  )
  expect_true(ggplot2::is.ggplot(p))
})

test_that("image geom with directory and extension options", {
  data <- mtcars %>%
    rownames_to_column("id") %>%
    arrange(desc(mpg)) %>%
    head(20) %>%
    mutate(type = c(rep("ice", 10), rep("electric", 10)))

  column_info <- tribble(
    ~id,     ~group,         ~name,                      ~geom,        ~palette,    ~options,
    "id",    NA,             "",                         "text",       NA,          list(hjust = 0, width = 6),
    "mpg",   NA,             "Miles / gallon",           "bar",        NA,          list(width = 4, legend = FALSE),
    "type",  NA,             "Type of engine",           "image",      NA,          lst(directory = "test_images/", extension = "png")
  )
  
  g <- funky_heatmap(
    data = data,
    column_info = column_info,
    expand = list(xmax = 4)
  )
  expect_true(ggplot2::is.ggplot(p))
})
```

Ideally, you could add more checks to ensure that indeed an image is being plotted with cowplot.

Reply:
I adapted the tests to this format, thanks! However, the use of `NA` caused this assert `is.character(column_info$palette) | is.factor(column_info$palette)` in `verify_column_info.R` to fail. I circumvented the issue by using `NA_character_`,  but maybe either this assert, or the if statement before this assert should be changed?

Reply:
Good point!

While I think `NA_character_` is the 'better' option here, I agree `NA` should also work.

I think it'd be best to add a statement before the assert that converts any NA to NA_character_. Or just add `is.na(column_info$palette) |` to the assert.